### PR TITLE
Whales - Uyen Vong

### PR DESCRIPTION
**Pull request by uvong**

### CODE REVIEW

**Comment by goeunpark** [comment]
Great job, Uyen! 🎉 

All your tests pass, the code comments and git commit messages are clear, your use of helper functions are 🔥 excellent 🔥 . Overall, this project is a Green 🟢 !

I've left a few comments in how you could make your code more readable and efficient! I noticed that you often use this syntax to loop within a nested structure (L29-L37, L68-L70, L94, L119, L130 in `party.py`):
```
for i in range(len(data_structure['key'])):
    # some logic here
``` 
Which works just fine -- but the syntax below would set you up to loop in the same position _and_ you can reference items by using `thing` (or whatever it is) instead of `data_structure['key'][i]`: 
```
things = data_structure['key']
for thing in things: 
    # some logic here
```
Let me know if you have any questions regarding my comments! Wonderful work, keep it up!

File: tests/test_wave_03.py [hunk 1]
```diff
@@ -58,8 +58,9 @@ def test_friends_unique_movies_not_duplicated():
     # *************************************************************************************************
     # ****** Add assertions here to test that the correct movies are in friends_unique_movies **********
     # **************************************************************************************************
+    assert INTRIGUE_3 in friends_unique_movies
```
Comment (goeunpark) [inline tests/test_wave_03.py:61]: I would like to see more assertions here to truly test that `friend_unique_movies` do not have  duplicates, as it describes to do so on L47. It's customary for our test definition to describe in full detail what our test block intends to test. Currently, this assertion only says that `INTRIGUE_3` is in `friends_unique_movies`, it doesn't actually test that `friends_unique_movies` has three distinct movies!

File: viewing_party/party.py [hunk 2]
```diff
@@ -1,23 +1,180 @@
 # ------------- WAVE 1 --------------------
-
+#adding a comment in the code to commit :)
 def create_movie(title, genre, rating):
-    pass
+    
+    user_data = {
+        "watched": [],
+        "watchlist": []
+        }
```
Comment (goeunpark) [inline viewing_party/party.py:8]: Great job looking ahead but we don't need or use this `user_data` variable in our function -- if it ain't used, delete it!

File: viewing_party/party.py [hunk 3]
```diff
@@ -1,23 +1,180 @@
...
+        "watchlist": []
+        }
+
+    new_movie = {
+        "title": title,
+        "genre": genre,
+        "rating": rating
+    }
+    if None in new_movie.values():
+        return None
```
Comment (goeunpark) [inline viewing_party/party.py:16]: Nice validation check!

Additionally, It's an even better idea to run checks like this before we take up memory with a possibly unnecessary variable -- we could move this check above setting your `new_movie` dictionary and return None if title or genre or movie == None.

File: viewing_party/party.py [hunk 4]
```diff
@@ -1,23 +1,180 @@
...
+    """compare the title argument to value of title in
+    each movie dictionary in the user's watchlist
+    if the titles are the same then append the movie dictionary
+    to the user's watched list and delete the dictionary
+    from the user's watchlist"""
+
+    for i in range(watchlist_length):
+        if user_data['watchlist'][i]['title'] == title:  
+            user_data['watched'].append(user_data['watchlist'][i])
+            del user_data['watchlist'][i]
```
Comment (goeunpark) [inline viewing_party/party.py:40]: Nice -- I haven't personally used the `del` keyword before! Between `.remove()` and `.pop()` and `del`, there's so many ways to remove items from a list (and for fun, I went down [this stack overflow thread](https://stackoverflow.com/questions/11520492/difference-between-del-remove-and-pop-on-lists) for the difference!!)

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,180 @@
...
+    
+    for movie in user_data['watched']:
+        sum_rating += movie["rating"]
 
+    avg_rating = sum_rating/watched_length
+    
+    return avg_rating
 
+def get_most_watched_genre(user_data):
+    popular_genre = None
```
Comment (goeunpark) [inline viewing_party/party.py:65]: I would consider grouping `popular_genre` and `popular_count` together because they are related OR combining that information as a dictionary!

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,180 @@
...
+    genre_list = []
+    popular_count = 0
+    watched_length = len(user_data['watched'])
+    
+    for i in range(watched_length):
+        genre_list.append(user_data['watched'][i]['genre'])
+
+    #could use Counter module but stuck to naive approach
```
Comment (goeunpark) [inline viewing_party/party.py:73]: I appreciate the naive approach because now I know that you know how to implement a counter in python! 💯

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,180 @@
...
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+
+    user_watched_titles = []
+    friends_movies = []
+    unique_movies = []
+    
+    #generate a list of title values that user watched
+    for movie in user_data['watched']:
```
Comment (goeunpark) [inline viewing_party/party.py:90]: ✨ Yes! This is the for loop syntax that makes the readers' life so much easier!

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,180 @@
...
+    
+    #generate a list of title values that friends watched
+    for i in range(len(user_data['friends'])):
+        for movie in user_data['friends'][i]['watched']:
+            friends_movies.append(movie['title'])
 
+    #convert to set to get unique titles via difference method
+    user_watched_set = set(user_watched_titles)
+    friends_movies_set = set(friends_movies)
+    unique_watched = user_watched_set - friends_movies_set
```
Comment (goeunpark) [inline viewing_party/party.py:101]: Great use of a set's core functionality here!

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,180 @@
...
+    user_watched_set = set(user_watched_titles)
+    friends_movies_set = set(friends_movies)
+    unique_watched = friends_movies_set - user_watched_set
+    
+    #avoiding duplicate movies by not adding movie that 
+    #is already in friends_unique_movies list
+
+    for i in range(len(user_data['friends'])):
+        for movie in user_data['friends'][i]['watched']:
+            if movie['title'] in unique_watched and movie not in friends_unique_movies:
```
Comment (goeunpark) [inline viewing_party/party.py:132]: Love checking two conditions in one if statement line!

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,180 @@
...
+                 friends_unique_movies.append(movie)
+    
+    return friends_unique_movies
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    movie_recs = []
+    friends_unique_watched = get_friends_unique_watched(user_data)
```
Comment (goeunpark) [inline viewing_party/party.py:142]: Great use of the `get_friends_unique_watched()` helper function!

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,180 @@
...
+    return movie_recs
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    movie_recs = []
+    user_watched = []
+    popular_genre = get_most_watched_genre(user_data)
+    friends_unique_watched = get_friends_unique_watched(user_data)
```
Comment (goeunpark) [inline viewing_party/party.py:157]: Wonderful, wonderful

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,180 @@
...
+    movie_recs = []
+    user_unique_movies = get_unique_watched(user_data)
+    user_unique_movie_titles = []
+
+    for movie in user_unique_movies:
+        user_unique_movie_titles.append(movie['title'])
+
+    for movie in user_data['favorites']:
+        if movie['title'] in user_unique_movie_titles:
+            movie_recs.append(movie)
```
Comment (goeunpark) [inline viewing_party/party.py:178]: We can also shorten this two for-loop logic into one for loop and get rid of the intermediate `user_unique_movie_titles` list, if desired:
```
for movie in user_unique_movies:
    if movie in user_data['favorites']:
        movie_recs.append(movie)
```
